### PR TITLE
Simplify c10::string_view

### PR DESCRIPTION
**Pull request by cyyever**
Make c10::basic_string_view a subclass of std::basic_string_view for easier replacement in the future.

### CODE REVIEW

**Review by zou3519 — CHANGES_REQUESTED** [review]
tests are failing

File: c10/test/util/string_view_test.cpp [hunk 1]
```diff
@@ -366,15 +330,6 @@ static_assert(string_view("hello").substr(0, 100) == string_view("hello"), "");
...
 
-TEST(StringViewTest, whenCallingSubstrWithPosOutOfRange_thenThrows) {
-  expectThrows<std::out_of_range>(
-      [] { string_view("hello").substr(6); },
-      "basic_string_view::substr parameter out of bounds. Index: 6, size: 5");
-
-  expectThrows<std::out_of_range>(
-      [] { string_view("hello").substr(6, 0); },
-      "basic_string_view::substr parameter out of bounds. Index: 6, size: 5");
-}
```
Comment (ezyang) [inline c10/test/util/string_view_test.cpp:377]: So this is what, standard string view has less error checking for out of bounds stuff?

File: c10/util/string_view.h [hunk 2]
```diff
@@ -15,579 +15,120 @@
...
-/**
- * Port of std::string_view with methods from C++20.
- * Implemented following the interface definition in
- * https://en.cppreference.com/w/cpp/string/basic_string_view
- * See there for the API documentation.
- *
- * Difference: We don't have a Traits template parameter because
- * std::char_traits isn't constexpr and we'd have to reimplement
- * std::char_traits if we wanted to use it with our constexpr basic_string_view.
- */
```
Comment (ezyang) [inline c10/util/string_view.h:27]: Can we have an updated comment for what this class does ontop of the std version?

**Review by ezyang — APPROVED** [review]

File: c10/test/util/string_view_test.cpp (hunk 1 above)
Comment (cyyever) [reply c10/test/util/string_view_test.cpp:377]: Different compilers throw different exception messages for std::out_of_range although the standard requires std::out_of_range.

Comment (ezyang) [reply c10/test/util/string_view_test.cpp:377]: ok. I guess it's not a big deal to stop testing the std class anyway

File: c10/util/string_view.h (hunk 2 above)
Comment (cyyever) [reply c10/util/string_view.h:27]: Updated

**Comment by izaitsevfb** [comment]
Looks like this change is BC-breaking: 500+ tests failed internally at meta after importing it.

```
TorchScript\NNTorchScriptBackend.cpp(630): error C2440: '<function-style-cast>': cannot convert from 'initializer list' to 'fmt::v9::string_view'
TorchScript\NNTorchScriptBackend.cpp(630): note: No constructor could take the source type, or constructor overload resolution was ambiguous
```

Reverting for now, will look deeper into how this change can be shephered in later.

**Comment by malfet** [comment]
@izaitsevfb this sounds like another dependency update problem, we need to keep version of third_party dependencies in sync between internal and external

**Comment by cyyever** [comment]
@izaitsevfb Can help update the internal libfmt?

**Comment by izaitsevfb** [comment]
> @izaitsevfb Can help update the internal libfmt?

I checked what that would entail, and unfortunately it's problematic: we're at v9.1.0 with one patch and multiple hacks on top of it.

@malfet, what are your thoughts on how we should proceed?

**Pull request closed without merging**